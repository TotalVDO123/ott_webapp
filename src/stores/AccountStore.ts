import { Store } from 'pullstate';
import jwtDecode from 'jwt-decode';

import * as accountService from '../services/account.service';
import * as subscriptionService from '../services/subscription.service';
import type { AuthData, Capture, Customer, JwtDetails, CustomerConsent, Consent } from '../../types/account';
import * as persist from '../utils/persist';
import type { PaymentDetail, Subscription, Transaction } from '../../types/subscription';
import { fetchCustomerConsents, fetchPublisherConsents, updateCustomer } from '../services/account.service';
import { getPaymentDetails, getTransactions } from '../services/subscription.service';

import { ConfigStore } from './ConfigStore';
import { watchHistoryStore, restoreWatchHistory, serializeWatchHistory } from './WatchHistoryStore';
import { favoritesStore, restoreFavorites, serializeFavorites } from './FavoritesStore';

const PERSIST_KEY_ACCOUNT = 'auth';

type AccountStore = {
  loading: boolean;
  auth: AuthData | null;
  user: Customer | null;
  subscription: Subscription | null;
  transactions: Transaction[] | null;
  activePayment: PaymentDetail | null;
  customerConsents: CustomerConsent[] | null;
  publisherConsents: Consent[] | null;
};

export const AccountStore = new Store<AccountStore>({
  loading: true,
  auth: null,
  user: null,
  subscription: null,
  transactions: null,
  activePayment: null,
  customerConsents: null,
  publisherConsents: null,
});

const setLoading = (loading: boolean) => {
  return AccountStore.update((s) => {
    s.loading = loading;
  });
};

let subscription: undefined | (() => void);
let refreshTimeout: number;

export const initializeAccount = async () => {
  const {
    config: { cleengId, cleengSandbox },
  } = ConfigStore.getRawState();

  if (!cleengId) {
    setLoading(false);
    return;
  }

  const storedSession: AuthData | null = persist.getItem(PERSIST_KEY_ACCOUNT) as AuthData | null;

  // clear previous subscribe (for dev environment only)
  if (subscription) {
    subscription();
  }

  subscription = AccountStore.subscribe(
    (state) => state.auth,
    (authData) => {
      window.clearTimeout(refreshTimeout);

      if (authData) {
        refreshTimeout = window.setTimeout(() => refreshJwtToken(cleengSandbox, authData), 60 * 1000);
      }

      persist.setItem(PERSIST_KEY_ACCOUNT, authData);
    },
  );

  // restore session from localStorage
  try {
    if (storedSession) {
      const refreshedAuthData = await getFreshJwtToken(cleengSandbox, storedSession);

      if (refreshedAuthData) {
        await afterLogin(cleengSandbox, refreshedAuthData);
        await restoreWatchHistory();
        await restoreFavorites();
      }
    }
  } catch (error: unknown) {
    await logout();
  }

  setLoading(false);
};

export async function updateUser(values: { firstName: string; lastName: string } | { email: string; confirmationPassword: string }) {
  const { auth, user } = AccountStore.getRawState();

  if (!auth || !user) throw new Error('no auth');

  const {
    config: { cleengSandbox },
  } = ConfigStore.getRawState();

  const response = await updateCustomer({ ...values, id: user.id.toString() }, cleengSandbox, auth.jwt);

  if (!response) {
    return { errors: Array.of('Unknown error') };
  }

  if (response.errors?.length === 0) {
    AccountStore.update((s) => {
      s.user = response.responseData;
    });
  }

  return response;
}

const getFreshJwtToken = async (sandbox: boolean, auth: AuthData) => {
  const result = await accountService.refreshToken({ refreshToken: auth.refreshToken }, sandbox);

  if (result.errors.length) throw new Error(result.errors[0]);

  return result?.responseData;
};

const refreshJwtToken = async (sandbox: boolean, auth: AuthData) => {
  try {
    const authData = await getFreshJwtToken(sandbox, auth);

    if (authData) {
      AccountStore.update((s) => {
        s.auth = { ...s.auth, ...authData };
      });
    }
  } catch (error: unknown) {
    // failed to refresh, logout user
    await logout();
  }
};

export const afterLogin = async (sandbox: boolean, auth: AuthData) => {
  const { accessModel } = ConfigStore.getRawState();
  const decodedToken: JwtDetails = jwtDecode(auth.jwt);
  const customerId = decodedToken.customerId;
  const response = await accountService.getCustomer({ customerId }, sandbox, auth.jwt);

  if (response.errors.length) throw new Error(response.errors[0]);

  AccountStore.update((s) => {
    s.auth = auth;
    s.user = response.responseData;
  });

  if (accessModel === 'SVOD') {
    await reloadActiveSubscription();
  }

  await getCustomerConsents();
  await getPublisherConsents();

  AccountStore.update((s) => {
    s.loading = false;
  });
};

export const login = async (email: string, password: string) => {
  await useConfig(async ({ cleengId, cleengSandbox }) => {
    setLoading(true);

    const response = await accountService.login({ email, password, publisherId: cleengId }, cleengSandbox);

    if (response.errors.length > 0) throw new Error(response.errors[0]);

    await afterLogin(cleengSandbox, response.responseData);

    await restoreFavorites();
    await restoreWatchHistory();
  });
};

export const logout = async () => {
  persist.removeItem(PERSIST_KEY_ACCOUNT);

  AccountStore.update((s) => {
    s.auth = null;
    s.user = null;
    s.customerConsents = null;
    s.publisherConsents = null;
  });

  await restoreFavorites();
  await restoreWatchHistory();
};

export const register = async (email: string, password: string) => {
  await useConfig(async ({ cleengId, cleengSandbox }) => {
    const localesResponse = await accountService.getLocales(cleengSandbox);

    if (localesResponse.errors.length > 0) throw new Error(localesResponse.errors[0]);

    const responseRegister = await accountService.register(
      {
        email: email,
        password: password,
        locale: localesResponse.responseData.locale,
        country: localesResponse.responseData.country,
        currency: localesResponse.responseData.currency,
        publisherId: cleengId,
      },
      cleengSandbox,
    );

    if (responseRegister.errors.length) throw new Error(responseRegister.errors[0]);

    await afterLogin(cleengSandbox, responseRegister.responseData);

    updatePersonalShelves();
  });
};

export const updatePersonalShelves = async () => {
  return await useLoginContext(async ({ cleengSandbox, customerId, auth: { jwt } }) => {
    const { watchHistory } = watchHistoryStore.getRawState();
    const { favorites } = favoritesStore.getRawState();

    if (!watchHistory && !favorites) return;

    const personalShelfData = { history: serializeWatchHistory(watchHistory), favorites: serializeFavorites(favorites) };

    return await accountService.updateCustomer(
      {
        id: customerId,
        externalData: personalShelfData,
      },
      cleengSandbox,
      jwt,
    );
  });
};

export const updateConsents = async (customerConsents: CustomerConsent[]) => {
  return await useLoginContext(async ({ cleengSandbox, customerId, auth: { jwt } }) => {
    const response = await accountService.updateCustomerConsents({ id: customerId, consents: customerConsents }, cleengSandbox, jwt);

    await getCustomerConsents();

    return response;
  });
};

export async function getCustomerConsents() {
  return await useLoginContext(async ({ cleengSandbox, customerId, auth: { jwt } }) => {
    const response = await fetchCustomerConsents({ customerId }, cleengSandbox, jwt);

    if (response && !response.errors?.length) {
      AccountStore.update((s) => {
        s.customerConsents = response.responseData.consents;
      });
    }

    return response;
  });
}

export async function getPublisherConsents() {
  return await useConfig(async ({ cleengId, cleengSandbox }) => {
    const response = await fetchPublisherConsents({ publisherId: cleengId }, cleengSandbox);

    if (response && !response.errors?.length) {
      AccountStore.update((s) => {
        s.publisherConsents = response.responseData.consents;
      });
    }

    return response;
  });
}

export const getCaptureStatus = async () => {
  return await useLoginContext(async ({ cleengSandbox, customerId, auth: { jwt } }) => {
    const response = await accountService.getCaptureStatus({ customerId }, cleengSandbox, jwt);

    if (response.errors.length > 0) throw new Error(response.errors[0]);

    return response.responseData;
  });
};

export const updateCaptureAnswers = async (capture: Capture) => {
  return await useLoginContext(async ({ cleengSandbox, customerId, auth }) => {
    const response = await accountService.updateCaptureAnswers({ customerId, ...capture }, cleengSandbox, auth.jwt);

    if (response.errors.length > 0) throw new Error(response.errors[0]);

    await afterLogin(cleengSandbox, auth);

    return response.responseData;
  });
};

export const resetPassword = async (email: string, resetUrl: string) => {
  return await useConfig(async ({ cleengId, cleengSandbox }) => {
    const response = await accountService.resetPassword(
      {
        customerEmail: email,
        publisherId: cleengId,
        resetUrl,
      },
      cleengSandbox,
    );

    if (response.errors.length > 0) throw new Error(response.errors[0]);

    return response.responseData;
  });
};

export const changePassword = async (customerEmail: string, newPassword: string, resetPasswordToken: string) => {
  return await useConfig(async ({ cleengId, cleengSandbox }) => {
    const response = await accountService.changePassword(
      {
        publisherId: cleengId,
        customerEmail,
        newPassword,
        resetPasswordToken,
      },
      cleengSandbox,
    );

    if (response.errors.length > 0) throw new Error(response.errors[0]);

    return response.responseData;
  });
};

export const updateSubscription = async (status: 'active' | 'cancelled') => {
  return await useLoginContext(async ({ cleengSandbox, customerId, auth: { jwt } }) => {
    const { subscription } = AccountStore.getRawState();
    if (!subscription) throw new Error('user has no active subscription');

    const response = await subscriptionService.updateSubscription(
      {
        customerId,
        offerId: subscription.offerId,
        status,
      },
      cleengSandbox,
      jwt,
    );

    if (response.errors.length > 0) throw new Error(response.errors[0]);

    await reloadActiveSubscription();

    return response.responseData;
  });
};

export async function reloadActiveSubscription({ delay }: { delay: number } = { delay: 0 }) {
  AccountStore.update((s) => {
    s.loading = true;
  });

  return await useLoginContext(async ({ cleengSandbox, customerId, auth: { jwt } }) => {
    const activeSubscription = await getActiveSubscription({ cleengSandbox, customerId, jwt });
    const transactions = await getAllTransactions({ cleengSandbox, customerId, jwt });
    const activePayment = await getActivePayment({ cleengSandbox, customerId, jwt });

    // The subscription data takes a few seconds to load after it's purchased,
    // so here's a delay mechanism to give it time to process
    if (delay > 0) {
      window.setTimeout(() => reloadActiveSubscription(), delay);
    } else {
      AccountStore.update((s) => {
        s.subscription = activeSubscription;
        s.transactions = transactions;
        s.activePayment = activePayment;
        s.loading = false;
      });
    }
  });
}

async function getActiveSubscription({ cleengSandbox, customerId, jwt }: { cleengSandbox: boolean; customerId: string; jwt: string }) {
  const response = await subscriptionService.getSubscriptions({ customerId }, cleengSandbox, jwt);

  if (response.errors.length > 0) return null;

  return response.responseData.items.find((item) => item.status === 'active' || item.status === 'cancelled') || null;
}

async function getAllTransactions({ cleengSandbox, customerId, jwt }: { cleengSandbox: boolean; customerId: string; jwt: string }) {
  const response = await getTransactions({ customerId }, cleengSandbox, jwt);

  if (response.errors.length > 0) return null;

  return response.responseData.items;
}

async function getActivePayment({ cleengSandbox, customerId, jwt }: { cleengSandbox: boolean; customerId: string; jwt: string }) {
  const response = await getPaymentDetails({ customerId }, cleengSandbox, jwt);

  if (response.errors.length > 0) return null;

  return response.responseData.paymentDetails.find((paymentDetails) => paymentDetails.active) || null;
}

function useConfig<T>(callback: (config: { cleengId: string; cleengSandbox: boolean }) => T): T {
  const {
    config: { cleengId, cleengSandbox },
  } = ConfigStore.getRawState();

  if (!cleengId) throw new Error('cleengId is not configured');

  return callback({ cleengId, cleengSandbox });
}

function useLoginContext<T>(callback: (args: { cleengId: string; cleengSandbox: boolean; customerId: string; auth: AuthData }) => T): T {
  const { user, auth } = AccountStore.getRawState();

  if (!user?.id || !auth?.jwt) throw new Error('user not logged in');

  return useConfig((config) => callback({ ...config, customerId: user.id, auth }));
}
