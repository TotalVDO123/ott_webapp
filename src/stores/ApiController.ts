import { injectable } from 'inversify';

import type { GetPlaylistParams, Playlist, PlaylistItem } from '#types/playlist';
import type { AdSchedule } from '#types/ad-schedule';
import type { EpisodeInSeries, EpisodesWithPagination, GetSeriesParams, Series } from '#types/series';
import ApiService from '#src/services/api.service';

@injectable()
export default class EpgController {
  private readonly apiService: ApiService;

  constructor(apiService: ApiService) {
    this.apiService = apiService;
  }

  getPlaylistById = async (id?: string, params: GetPlaylistParams = {}): Promise<Playlist | undefined> => {
    return this.apiService.getPlaylistById(id, params);
  };

  getMediaByWatchlist = async (playlistId: string, mediaIds: string[], token?: string): Promise<PlaylistItem[] | undefined> => {
    return this.apiService.getMediaByWatchlist(playlistId, mediaIds, token);
  };

  getMediaById = async (id: string, token?: string, drmPolicyId?: string): Promise<PlaylistItem | undefined> => {
    return this.apiService.getMediaById(id, token, drmPolicyId);
  };

  getSeries = async (id: string, params: GetSeriesParams = {}): Promise<Series | undefined> => {
    return this.apiService.getSeries(id, params);
  };

  getSeriesByMediaIds = async (mediaIds: string[]): Promise<{ [mediaId: string]: EpisodeInSeries[] | undefined } | undefined> => {
    return this.apiService.getSeriesByMediaIds(mediaIds);
  };

  getEpisodes = async (args: { seriesId: string | undefined; pageOffset?: number; pageLimit?: number; afterId?: string }): Promise<EpisodesWithPagination> => {
    return this.apiService.getEpisodes(args);
  };

  getSeasonWithEpisodes = async (args: {
    seriesId: string | undefined;
    seasonNumber: number;
    pageOffset?: number;
    pageLimit?: number;
  }): Promise<EpisodesWithPagination> => {
    return this.apiService.getSeasonWithEpisodes(args);
  };

  getAdSchedule = async (id: string | undefined | null): Promise<AdSchedule | undefined> => {
    return this.apiService.getAdSchedule(id);
  };
}
