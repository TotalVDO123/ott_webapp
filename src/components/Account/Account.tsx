import React, { useState, useMemo } from 'react';
import { useTranslation } from 'react-i18next';
import type { GenericFormValues, CustomerFormErrors } from 'types/form';
import { useHistory } from 'react-router-dom';

import { formatConsentsFromValues, formatConsentValues } from '../../utils/collection';
import Visibility from '../../icons/Visibility';
import VisibilityOff from '../../icons/VisibilityOff';
import useToggle from '../../hooks/useToggle';
import Button from '../../components/Button/Button';
import Form from '../Form/Form';
import IconButton from '../IconButton/IconButton';
import LoadingOverlay from '../LoadingOverlay/LoadingOverlay';
import TextField from '../TextField/TextField';
import Checkbox from '../Checkbox/Checkbox';
import { addQueryParam } from '../../utils/history';
import { AccountStore, updateConsents, updateUser } from '../../stores/AccountStore';

import styles from './Account.module.scss';

type Props = {
  panelClassName?: string;
  panelHeaderClassName?: string;
};

type Editing = 'none' | 'account' | 'password' | 'info' | 'consents';

const Account = ({ panelClassName, panelHeaderClassName }: Props): JSX.Element => {
  const { t } = useTranslation('user');
  const history = useHistory();
  const [editing, setEditing] = useState<Editing>('none');
  const [errors, setErrors] = useState<CustomerFormErrors | undefined>();
  const [isLoading, setIsLoading] = useState(false);
  const [viewPassword, toggleViewPassword] = useToggle();

  const { user: customer, customerConsents, publisherConsents } = AccountStore.useState((state) => state);
  const consentValues = useMemo(() => formatConsentValues(publisherConsents, customerConsents), [publisherConsents, customerConsents]);
  const initialValues = useMemo(() => ({ ...customer, consents: consentValues }), [customer, consentValues]);

  const formatConsentLabel = (label: string): string | JSX.Element => {
    // @todo sanitize consent label to prevent XSS
    const hasHrefOpenTag = /<a(.|\n)*?>/.test(label);
    const hasHrefCloseTag = /<\/a(.|\n)*?>/.test(label);

    if (hasHrefOpenTag && hasHrefCloseTag) {
      return <span dangerouslySetInnerHTML={{ __html: label }} />;
    }

    return label;
  };

  const translateErrors = (errors?: string[]) => {
    const formErrors: CustomerFormErrors = {};

    // Some errors are combined in a single CSV string instead of one string per error
    errors
      ?.flatMap((e) => e.split(','))
      .map((error) => {
        switch (error.trim()) {
          case 'Invalid param email':
            formErrors.email = t('account.errors.invalid_param_email');
            break;
          case 'Customer email already exists':
            formErrors.email = t('account.errors.email_exists');
            break;
          case 'Please enter a valid e-mail address.':
            formErrors.email = t('account.errors.please_enter_valid_email');
            break;
          case 'Invalid confirmationPassword': {
            formErrors.confirmationPassword = t('account.errors.invalid_password');
            break;
          }
          case 'firstName can have max 50 characters.': {
            formErrors.firstName = t('account.errors.first_name_too_long');
            break;
          }
          case 'lastName can have max 50 characters.': {
            formErrors.lastName = t('account.errors.last_name_too_long');
            break;
          }
          default:
            console.info('Unknown error', error);
            return;
        }
      });
    return formErrors;
  };

  async function handleSubmit(values: GenericFormValues) {
    let response: ApiResponse | undefined = undefined;
    setIsLoading(true);
    switch (editing) {
      case 'account':
        response = await updateUser({ email: values.email, confirmationPassword: values.confirmationPassword });
        break;
      case 'info':
        response = await updateUser({ firstName: values.firstName, lastName: values.lastName });
        break;
      case 'consents':
        response = await updateConsents(formatConsentsFromValues(publisherConsents, values));
        break;
      default:
        return;
    }

    setErrors(translateErrors(response?.errors));

    if (response && !response?.errors?.length) {
      setEditing('none');
    }

    setIsLoading(false);
  }

  const onCancelClick = (formResetHandler?: () => void): void => {
    formResetHandler && formResetHandler();
    setErrors(undefined);
    setEditing('none');
  };

  const editPasswordClickHandler = () => {
    history.push(addQueryParam(history, 'u', 'reset-password'));
  };

  return (
    <Form initialValues={initialValues} onSubmit={handleSubmit}>
      {({ values, handleChange, handleReset, handleSubmit, hasChanged }) => (
        <>
          {isLoading && <LoadingOverlay transparentBackground />}
          <div className={panelClassName}>
            <div className={panelHeaderClassName}>
              <h3>{t('account.email')}</h3>
            </div>
            <div className={styles.flexBox}>
              <TextField
                name="email"
                label={t('account.email')}
                value={values.email as string}
                onChange={handleChange}
                error={!!errors?.email}
                helperText={errors?.email}
                disabled={isLoading}
                editing={editing === 'account'}
                required
              />
              {editing === 'account' && (
                <TextField
                  name="confirmationPassword"
                  label={t('account.confirm_password')}
                  value={values.confirmationPassword as string}
                  onChange={handleChange}
                  error={!!errors?.confirmationPassword}
                  helperText={errors?.confirmationPassword}
                  type={viewPassword ? 'text' : 'password'}
                  disabled={isLoading}
                  rightControl={
                    <IconButton aria-label={viewPassword ? t('account.hide_password') : t('account.view_password')} onClick={() => toggleViewPassword()}>
                      {viewPassword ? <Visibility /> : <VisibilityOff />}
                    </IconButton>
                  }
                  required
                />
              )}
              <div className={styles.controls}>
                {editing === 'account' ? (
                  <>
                    <Button
                      label={t('account.save')}
                      type="submit"
                      onClick={handleSubmit}
                      disabled={isLoading || !values.email || !values.confirmationPassword}
                    />
                    <Button label={t('account.cancel')} type="reset" variant="text" onClick={() => onCancelClick(handleReset)} />
                  </>
                ) : (
                  <Button label={t('account.edit_account')} type="button" onClick={() => setEditing('account')} />
                )}
              </div>
            </div>
          </div>
          <div className={panelClassName}>
            <div className={panelHeaderClassName}>
              <h3>{t('account.security')}</h3>
            </div>
            <div>
              <strong>{t('account.password')}</strong>
              <p>****************</p>
              <Button label={t('account.edit_password')} type="button" onClick={() => (customer ? editPasswordClickHandler() : null)} />
            </div>
          </div>
          <div className={panelClassName}>
            <div className={panelHeaderClassName}>
              <h3>{t('account.about_you')}</h3>
            </div>
            <div>
              <div className={styles.flexBox}>
                <TextField
                  name="firstName"
                  label={t('account.firstname')}
                  value={values.firstName as string}
                  onChange={handleChange}
                  error={!!errors?.firstName}
                  helperText={errors?.firstName}
                  disabled={isLoading}
                  editing={editing === 'info'}
                />
                <TextField
                  name="lastName"
                  label={t('account.lastname')}
                  value={values.lastName as string}
                  onChange={handleChange}
                  error={!!errors?.lastName}
                  helperText={errors?.lastName}
                  disabled={isLoading}
                  editing={editing === 'info'}
                />
                <div className={styles.controls}>
                  {editing === 'info' ? (
                    <>
                      <Button label={t('account.save')} type="submit" disabled={!hasChanged} onClick={handleSubmit} />
                      <Button type="reset" variant="text" label={t('account.cancel')} onClick={() => setEditing('none')} />
                    </>
                  ) : (
                    <Button type="button" label={t('account.edit_information')} onClick={() => setEditing('info')} />
                  )}
                </div>
              </div>
            </div>
          </div>
          {publisherConsents && (
            <div className={panelClassName}>
              <div className={panelHeaderClassName}>
                <h3>{t('account.terms_and_tracking')}</h3>
              </div>
              <div className={styles.flexBox} onClick={() => setEditing('consents')}>
                {publisherConsents.map((consent, index) => (
                  <Checkbox
                    key={index}
                    name={consent.name}
                    value={consent.value || ''}
                    checked={(values.consents?.[consent.name] as boolean) || false}
                    onChange={(event) => (handleChange ? handleChange(event, { nestInto: 'consents' }) : null)}
                    label={formatConsentLabel(consent.label)}
                    disabled={consent.required}
                  />
                ))}
                <div className={styles.controls}>
                  <Button
                    id="submit_consents"
                    className={styles.submitConsents}
                    type="button"
                    label={t('account.update_consents')}
                    disabled={!hasChanged}
                    onClick={handleSubmit}
                  />
                </div>
              </div>
            </div>
          )}
        </>
      )}
    </Form>
  );
};

export default Account;
