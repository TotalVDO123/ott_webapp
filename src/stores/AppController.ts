import merge from 'lodash.merge';
import { inject, injectable } from 'inversify';

import AccountController from './AccountController';
import WatchHistoryController from './WatchHistoryController';
import FavoritesController from './FavoritesController';

import type { Config } from '#types/Config';
import { IntegrationInfo, useConfigStore } from '#src/stores/ConfigStore';
import ConfigService from '#src/services/config.service';
import SettingsService from '#src/services/settings.service';
import { PersonalShelf } from '#src/config';
import { ConfigError } from '#src/utils/error';
import { getModule } from '#src/container';
import { logDev } from '#src/utils/common';

type IntegrationRegistration = {
  name: string;
  selector: (config: Config) => boolean;
  register: (config: Config) => Promise<void> | void;
};

@injectable()
export default class AppController {
  private readonly configService: ConfigService;
  private readonly settingsService: SettingsService;

  private integrationRegistrations: IntegrationRegistration[] = [];

  constructor(@inject(ConfigService) configService: ConfigService, @inject(SettingsService) settingsService: SettingsService) {
    this.configService = configService;
    this.settingsService = settingsService;
  }

  registerIntegration(registration: IntegrationRegistration) {
    this.integrationRegistrations.push(registration);
  }

  async loadIntegration(config: Config) {
    const registration = this.integrationRegistrations.find(({ selector }) => selector(config));

    if (registration) {
      logDev(`Found integration: ${registration.name}`);
      await registration.register(config);
    }
  }

  loadAndValidateConfig = async (configSource: string | undefined) => {
    const configLocation = this.configService.formatSourceLocation(configSource);
    const defaultConfig = this.configService.getDefaultConfig();
    let config: Config | null = null;

    if (!configLocation) {
      useConfigStore.setState({ config: defaultConfig });
      throw new ConfigError('Config not defined');
    }

    try {
      config = await this.configService.loadConfig(configLocation);
    } catch (err: unknown) {
      throw new ConfigError('Config not found');
    }

    if (!config) {
      throw new ConfigError('Config not found');
    }

    config.id = configSource;
    config.assets = config.assets || {};

    // make sure the banner always defaults to the JWP banner when not defined in the config
    if (!config.assets.banner) {
      config.assets.banner = defaultConfig.assets.banner;
    }

    // Store the logo right away and set css variables so the error page will be branded
    const banner = config.assets.banner;

    useConfigStore.setState((s) => {
      s.config.assets.banner = banner;
    });

    this.configService.setCssVariables(config.styling || {});

    config = await this.configService.validateConfig(config);
    config = merge({}, defaultConfig, config);

    const accessModel = this.configService.calculateAccessModel(config);

    useConfigStore.setState({ config, accessModel });

    this.configService.maybeInjectAnalyticsLibrary(config);

    if (config.adSchedule) {
      const adScheduleData = await this.configService.loadAdSchedule(config?.adSchedule);
      useConfigStore.setState({ adScheduleData });
    }

    return config;
  };

  initApp = async () => {
    const settings = await this.settingsService.initialize();
    const configSource = this.settingsService.getConfigSource(settings);
    const config = await this.loadAndValidateConfig(configSource);

    // update settings in the config store
    useConfigStore.setState({ settings });

    // load the integration based on the app config
    await this.loadIntegration(config);

    if (config.features?.continueWatchingList && config.content.some((el) => el.type === PersonalShelf.ContinueWatching)) {
      await getModule(WatchHistoryController).restoreWatchHistory();
    }

    if (config.features?.favoritesList && config.content.some((el) => el.type === PersonalShelf.Favorites)) {
      await getModule(FavoritesController).initializeFavorites();
    }

    if (this.getIntegration().integrationType) {
      await getModule(AccountController).initializeAccount();
    }

    return { config, settings, configSource };
  };

  getIntegration = (): IntegrationInfo => {
    return useConfigStore.getState().getIntegration();
  };
}
