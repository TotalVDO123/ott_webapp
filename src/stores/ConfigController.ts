import merge from 'lodash.merge';
import { injectable } from 'inversify';

import { IntegrationInfo, useConfigStore } from '#src/stores/ConfigStore';
import ConfigService from '#src/services/config.service';

@injectable()
export default class ConfigController {
  private readonly configService: ConfigService;

  constructor(configService: ConfigService) {
    this.configService = configService;
  }

  initializeAdSchedule = async () => {
    const { config } = useConfigStore.getState();

    const adScheduleData = await this.configService.loadAdSchedule(config?.adSchedule);

    useConfigStore.setState({
      adScheduleData,
    });
  };

  loadAndValidateConfig = async (configSource: string | undefined) => {
    const configLocation = this.configService.formatSourceLocation(configSource);
    const defaultConfig = this.configService.getDefaultConfig();

    if (!configLocation) {
      useConfigStore.setState({ config: defaultConfig });
      throw new Error('Config not defined');
    }

    let config = await this.configService.loadConfig(configLocation);

    if (!config) {
      throw new Error('Config not found');
    }

    config.id = configSource;
    config.assets = config.assets || {};

    // make sure the banner always defaults to the JWP banner when not defined in the config
    if (!config.assets.banner) {
      config.assets.banner = defaultConfig.assets.banner;
    }

    // Store the logo right away and set css variables so the error page will be branded
    useConfigStore.setState((s) => {
      s.config.assets.banner = config.assets.banner;
    });

    this.configService.setCssVariables(config.styling || {});

    config = await this.configService.validateConfig(config);
    config = merge({}, defaultConfig, config);

    const accessModel = this.configService.calculateAccessModel(config);

    useConfigStore.setState({ config, accessModel });

    this.configService.maybeInjectAnalyticsLibrary(config);

    if (config.adSchedule) {
      await this.initializeAdSchedule();
    }

    return config;
  };

  getIntegration = (): IntegrationInfo => {
    return useConfigStore.getState().getIntegration();
  };
}
