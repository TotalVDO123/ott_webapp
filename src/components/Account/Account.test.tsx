import React from 'react';

import customer from '../../fixtures/customer.json';
import { useAccountStore } from '../../stores/AccountStore';
import { renderWithRouter } from '../../../test/testUtils';

import Account from './Account';

import type { Consent } from '#types/account';

describe('<Account>', () => {
  test('renders and matches snapshot', () => {
    useAccountStore.setState({
      user: customer,
      publisherConsents: Array.of({ name: 'marketing', label: 'Receive Marketing Emails' } as Consent),
    });

    const { container } = renderWithRouter(<Account panelClassName={'panel-class'} panelHeaderClassName={'header-class'} />);

    // todo
    expect(container).toMatchSnapshot();
  });
});
