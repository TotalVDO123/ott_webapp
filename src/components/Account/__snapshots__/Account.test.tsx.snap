// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Account> renders and matches snapshot 1`] = `
<div>
  <form
    novalidate=""
  >
    <div
      class="panel-class"
    >
      <div
        class="header-class"
      >
        <h3>
          account.email
        </h3>
      </div>
      <div
        class="flexBox"
      >
        <div
          class="textField"
        >
          <label
            class="label"
            for="text-field_1235_email"
          >
            account.email
          </label>
          <p>
            email@domain.com
          </p>
        </div>
        <div
          class="controls"
        >
          <button
            class="button default outlined"
            type="button"
          >
            <span>
              account.edit_account
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      class="panel-class"
    >
      <div
        class="header-class"
      >
        <h3>
          account.security
        </h3>
      </div>
      <div>
        <strong>
          account.password
        </strong>
        <p>
          ****************
        </p>
        <button
          class="button default outlined"
          type="button"
        >
          <span>
            account.edit_password
          </span>
        </button>
      </div>
    </div>
    <div
      class="panel-class"
    >
      <div
        class="header-class"
      >
        <h3>
          account.about_you
        </h3>
      </div>
      <div>
        <div
          class="flexBox"
        >
          <div
            class="textField"
          >
            <label
              class="label"
              for="text-field_1235_firstName"
            >
              account.firstname
            </label>
            <p>
              John
            </p>
          </div>
          <div
            class="textField"
          >
            <label
              class="label"
              for="text-field_1235_lastName"
            >
              account.lastname
            </label>
            <p>
              Doe
            </p>
          </div>
          <div
            class="controls"
          >
            <button
              class="button default outlined"
              type="button"
            >
              <span>
                account.edit_information
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="panel-class"
    >
      <div
        class="header-class"
      >
        <h3>
          account.terms_and_tracking
        </h3>
      </div>
      <div
        class="flexBox"
      >
        <div
          class="checkbox"
        >
          <div
            class="row"
          >
            <input
              id="check-box_1235_marketing"
              name="marketing"
              type="checkbox"
              value=""
            />
            <label
              for="check-box_1235_marketing"
            >
              
              Receive Marketing Emails
            </label>
          </div>
        </div>
        <div
          class="controls"
        >
          <button
            aria-disabled="true"
            class="button submitConsents default outlined disabled"
            disabled=""
            id="submit_consents"
            type="button"
          >
            <span>
              account.update_consents
            </span>
          </button>
        </div>
      </div>
    </div>
  </form>
</div>
`;
